persistence/remote/src/main/java/org/infinispan/persistence/remote/RemoteStore.java

```diff
@@ -68,228 +72,257 @@
+         remoteCacheManager = new RemoteCacheManager(builder.build());
+
+         if (configuration.remoteCacheName().isEmpty())
+            remoteCache = (InternalRemoteCache<Object, Object>) remoteCacheManager.getCache();
+         else
+            remoteCache = (InternalRemoteCache<Object, Object>) remoteCacheManager.getCache(configuration.remoteCacheName());
+
+         return remoteCache.ping();
+      }, "RemoteStore-start")
+            .thenCompose(Function.identity())
+            .thenAccept(pingResponse -> {
+               String cacheName = ctx.getCache().getName();
+               StorageConfigurationManager storageConfigurationManager = ctx.getCache().getAdvancedCache().getComponentRegistry()
+                     .getComponent(StorageConfigurationManager.class);
+               MediaType localKeyStorageType = storageConfigurationManager.getKeyStorageMediaType();
+               MediaType serverKeyStorageType = pingResponse.getKeyMediaType();
+
+               MediaType localValueStorageType = storageConfigurationManager.getValueStorageMediaType();
+               MediaType serverValueStorageType = pingResponse.getValueMediaType();
+
+               DataFormat.Builder dataFormatBuilder = DataFormat.builder().from(remoteCache.getDataFormat());
+               Integer numSegments = remoteCache.getCacheTopologyInfo().getNumSegments();
+               boolean segmentsMatch;
+               if (numSegments == null) {
+                  log.debugf("Remote Store for cache %s cannot support segmentation as the number of segments was not found from the remote cache", cacheName);
+                  segmentsMatch = false;
+               } else {
+                  segmentsMatch = numSegments == segmentCount;
+                  if (segmentsMatch) {
+                     log.debugf("Remote Store for cache %s can support segmentation as the number of segments matched the remote cache", cacheName);
+                  } else {
+                     log.debugf("Remote Store for cache %s cannot support segmentation as the number of segments %d do not match the remote cache %d",
+                           cacheName, segmentCount, numSegments);
+                  }
+               }
+               if (!segmentsMatch && configuration.segmented()) {
+                  throw log.segmentationRequiresEqualSegments(segmentCount, numSegments);
+               }
+               if ((supportsSegmentation = localKeyStorageType.equals(serverKeyStorageType))) {
+                  if (localKeyStorageType.isBinary()) {
+                     dataFormatBuilder.keyType(localKeyStorageType);
+                     // Make sure when the media types match that we don't marshall the value
+                     dataFormatBuilder.keyMarshaller(IdentityMarshaller.INSTANCE);
+                  } else {
+                     dataFormatBuilder.keyType(marshaller.mediaType());
+                     dataFormatBuilder.keyMarshaller(marshaller);
+                  }
```

Comment:
We can reduce the verbosity of this and keep the logic in one place by adding a couple of methods that are then used for setting both key/value types and marshaller:

```java
               if ((supportsSegmentation = localKeyStorageType.equals(serverKeyStorageType))) {
                  dataFormatBuilder.keyType(getMediaType(localKeyStorageType, marshaller));
                  dataFormatBuilder.keyMarshaller(getMarshaller(localKeyStorageType, marshaller));
               } else if (configuration.segmented()) {
                  throw log.segmentationRequiresEqualMediaTypes(localKeyStorageType, serverKeyStorageType);
               }
               // We store MarshalledEntry instances for the values when raw values is not set, instead of the stored value
               if (!configuration.rawValues()) {
                  localValueStorageType = MediaType.APPLICATION_OBJECT;
               }
               if (localValueStorageType.equals(serverValueStorageType)) {
                  dataFormatBuilder.valueType(getMediaType(localKeyStorageType, marshaller));
                  dataFormatBuilder.valueMarshaller(getMarshaller(localKeyStorageType, marshaller));
               }
...
   private MediaType getMediaType(MediaType local, Marshaller marshaller) {
      return local.isBinary() ? local : marshaller.mediaType();
   }

   private Marshaller getMarshaller(MediaType local, Marshaller marshaller) {
      return local.isBinary() ? IdentityMarshaller.INSTANCE : marshaller;
   }
```

Or we could just have the ternary operator inline.

Reply:
Hrmm, sure I can do this.

Reply:
Unfortunately, this leads to a bit of a weird logic in that getMediaType takes a marshaller that is itself dependent upon getMarshaller as we don't want to ever use the media type from IdentityMarshaller. I will play this a bit more.

Reply:
I have changed it to ternary, which while doesn't consolidate the 2 branches think it makes the code a bit more straight forward that they are dependent upon each other.